src/globus_cli/commands/cli_profile_list.py

```diff
@@ -0,0 +1,97 @@
+import os
+from typing import Any, Dict, Iterator, List, Optional, Tuple
+
+import click
+import globus_sdk.tokenstorage
+
+from globus_cli.login_manager import token_storage_adapter
+from globus_cli.parsing import command
+from globus_cli.termio import FORMAT_TEXT_TABLE, formatted_print
+from globus_cli.types import FIELD_LIST_T
+
+
+# TODO: upstream this into the SDK as a method of the SQLiteStorageAdapter
+def _iter_namespaces(adapter: globus_sdk.tokenstorage.SQLiteAdapter) -> Iterator[str]:
+    conn = adapter._connection
+
+    cursor = conn.execute("SELECT DISTINCT namespace FROM token_storage;")
+    for row in cursor:
+        yield row[0]
+
+
+def _profilestr_to_datadict(s: str) -> Optional[Dict[str, Any]]:
+    if s.count("/") < 1:
+        return None
+    if s.count("/") < 2:
+        category, env = s.split("/")
+        if category == "clientprofile":  # should not be possible
+            return None
+        return {"client": False, "env": env, "profile": None, "default": True}
+    else:
+        category, env, profile = s.split("/", 2)
+        return {
+            "client": category == "clientprofile",
+            "env": env,
+            "profile": profile,
+            "default": False,
+        }
+
+
+def _parse_and_filter_profiles(
+    all: bool,
+) -> Tuple[List[Dict[str, Any]], List[Dict[str, Any]]]:
+    namespaces = list(_iter_namespaces(token_storage_adapter()))
+    globus_env = os.getenv("GLOBUS_SDK_ENVIRONMENT", "production")
+
+    client_profiles = []
+    user_profiles = []
+    for n in namespaces:
+        data = _profilestr_to_datadict(n)
+        if not data:  # skip any parse failures
+            continue
+        if (
+            data["env"] != globus_env and not all
+        ):  # unless --all was passed, skip other envs
+            continue
+        if data["client"]:
+            client_profiles.append(data)
+        else:
+            if all or data["profile"] is not None:
```

Comment:
I feel like we should include the default userprofile for the current environment even without all set. It feels odd/potentially confusing not having that profile shown even when its the active profile,

Reply:
In my first draft, I had a warning about the default profile not being listed (but that looked more confusing, so I cut it).

Let me look again at including the default profile again. With an "active/current" marker, it might work out nicer.